gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java

```diff
@@ -0,0 +1,197 @@
+    verify(unaryCallable, times(1)).futureCall(any(LabeledIntList.class));
+  }
+
+  /** Verifies exception occurred at RPC is propagated to element results */
+  @Test(expected = ExecutionException.class)
+  public void testBatchingFailed() throws Exception {
+    final Exception exception = new RuntimeException();
+    when(mockDescriptor.newRequestBuilder(any(LabeledIntList.class)))
+        .thenReturn(SQUARER_BATCHING_DESC_V2.newRequestBuilder(labeledIntList));
+    when(unaryCallable.futureCall(any(LabeledIntList.class)))
+        .thenReturn(ApiFutures.<List<Integer>>immediateFailedFuture(exception));
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) {
+                List<SettableApiFuture<Integer>> apiFutures = invocation.getArgument(1);
+                apiFutures.get(0).setException(exception);
+                return null;
+              }
+            })
+        .when(mockDescriptor)
+        .splitException(any(Throwable.class), Mockito.<SettableApiFuture<Integer>>anyList());
+    ApiFuture<Integer> failedResult = underTest.add(5);
+    underTest.flush();
+    verify(mockDescriptor, times(1)).newRequestBuilder(labeledIntList);
+
+    verify(mockDescriptor, times(1))
+        .splitException(any(Throwable.class), Mockito.<SettableApiFuture<Integer>>anyList());
+    verify(unaryCallable, times(1)).futureCall(any(LabeledIntList.class));
+    assertThat(failedResult.isDone()).isTrue();
+    try {
+      failedResult.get();
+    } catch (InterruptedException | ExecutionException e) {
+      Truth.assertThat(e.getCause()).isInstanceOf(RuntimeException.class);
+      throw e;
+    }
+  }
+
+  /** Tests accumulated element are resolved when {@link Batcher#flush()} is called. */
+  @Test
+  public void testBatchingWithCallable() throws Exception {
+    underTest = newBatcherInstance();
+    int limit = 100;
+    int batch = 10;
+    List<ApiFuture<Integer>> resultList = new ArrayList<>(limit);
+    for (int i = 0; i <= limit; i++) {
+      resultList.add(underTest.add(i));
+      if (i % batch == 0) {
+        underTest.flush();
+        for (int j = i - batch; j >= 0 && j < i; j++) {
+          Truth.assertThat(resultList.get(j).isDone()).isTrue();
+          Truth.assertThat(resultList.get(j).get()).isEqualTo(j * j);
+        }
+      }
+    }
+  }
+
+  /** Element results are resolved after batch is closed. */
+  @Test
+  public void testBatcherClose() throws Exception {
+    ApiFuture<Integer> result;
+    try (Batcher<Integer, Integer> batcher = newBatcherInstance()) {
+      result = batcher.add(5);
+    }
+    assertThat(result.isDone()).isTrue();
+    assertThat(result.get()).isEqualTo(25);
+  }
+
+  /** Validates exception when batch is called after {@link Batcher#close()}. */
+  @Test(expected = IllegalStateException.class)
+  public void testWhenNoElementAdded() throws Exception {
+    underTest.close();
+    underTest.add(1);
```

Comment:
Please avoid using expected exception annotations, especially with such generic exceptions....there is nothing to prove that we got to add()